rfcs/0031-devices-api.md

```diff
@@ -0,0 +1,363 @@
+    TVMDeviceWooflesActivate(device0); // Could reserve or enable certain circuitry ahead of time
+    TVMDeviceWooflesActivate(device1); // Could reserve or enable certain circuitry ahead of time
+
+    TVMDeviceWooflesOpen(device0); // Opens resource for use
+    operator(device0); // Pass resource_handle to operator
+    TVMDeviceWooflesClose(device0); // Close device use
+
+    TVMDeviceWooflesOpen(device1); // Opens resource for use
+    operator(device1); // Pass resource_handle to operator
+    TVMDeviceWooflesClose(device1); // Close device use
+
+    TVMDeviceWooflesDeactivate(device0); // Turn off the device
+    TVMDeviceWooflesDeactivate(device1); // Turn off the device
+}
+```
+
+This is a simple and likely sufficient set of hooks which can be used to manage these device transactions.
+
+## Device API Functions
+In the example of Zephyr, devices are already a first class concept so many of the functions will no-op but should synchronisation be required, an example implementation could be:
+
+```c
+#include <device.h>
+
+typedef struct {
+    struct device* dev;
+    k_mutex lock;
+} tvm_device_woofles_t;
+
+int32_t TVMDeviceWooflesInit(tvm_device_woofles_t* tvm_dev, struct device* zephyr_dev) {
+    k_mutex_init(&tvm_dev->lock);
+    tvm_dev->dev = zephyr_dev;
+}
+
+int32_t TVMDeviceWooflesActivate(tvm_device_woofles_t* tvm_dev) {}
+
+int32_t TVMDeviceWooflesOpen(tvm_device_woofles_t* tvm_dev) {
+    k_mutex_lock(&tvm_dev->lock, K_FOREVER);
+}
+
+int32_t TVMDeviceWooflesClose(tvm_device_woofles_t* tvm_dev) {
+    k_mutex_unlock(&tvm_dev->lock);
+}
+
+int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev) {}
+
+int32_t TVMDeviceWooflesDestroy(tvm_device_woofles_t* tvm_dev) {
+    tvm_dev->dev = NULL;
+}
+```
+
+Whereas for CMSIS, you can use the platform-specific function to encapsulate the API to our imaginary UART accessed accelerator:
+
+```c
+typedef struct {
+    ARM_DRIVER_USART* dev;
+} tvm_device_uart_accel_t;
+
+int32_t TVMDeviceUartAccelInit(tvm_device_uart_accel_t* tvm_dev, ARM_DRIVER_USART* uart_dev) {
+    uart_dev->Initialize(NULL);
+    tvm_dev->dev = uart_dev;
+}
+
+int32_t TVMDeviceUartAccelActivate(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelOpen(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelClose(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelDeactivate(tvm_device_uart_accel_t* tvm_dev) {}
+
+int32_t TVMDeviceUartAccelDestroy(tvm_device_uart_accel_t* tvm_dev) {
+    tvm_dev->dev->Uninitialize();
+}
+```
+
+## Operator Usage
+Each operator would be expected to utilise one device structure and be passed that as the `resource_handle` parameter, making the assumption that each operator or variant of an operator is only bound to one device at a time. In the following example it can be seen how a accelerators interface is implemented per platform to take this void pointer and call the platform specific driver code.
```

Comment:
could you specify how TVM knows which device structure an operator uses? also, can you qualify this is just for the C runtime and c++ runtime when BYOC targets are linked in?

Reply:
I'd rather avoid using the BYOC branding and sticking to the technical terms given BYOC is fairly heavily overloaded, I've added a few more words here to illustrate the `Target` configuration of this.